File: dsa/10+11.graph/257_Binary_Tree_Paths.java
package graph;
import java.util.*;

//  * Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> paths = new ArrayList<String>();
        this.treePathHelper(root, paths, new StringBuilder());
        return paths;
    }

    public void treePathHelper(TreeNode root, List<String> paths, StringBuilder prefixPath){
        if (root == null){
            return;
        }

        int prefixLength = prefixPath.length();
        if (prefixLength != 0){
            prefixPath.append("->");
        }
        prefixPath.append(String.valueOf(root.val));
        if (root.left == null && root.right == null){
            paths.add(prefixPath.toString());
        }
        else{
            this.treePathHelper(root.left, paths, prefixPath);
            this.treePathHelper(root.right, paths, prefixPath);
        }

        prefixPath.setLength(prefixLength);
    }
}

File: dsa/10+11.graph/257_Binary_Tree_Paths.py
import typing as t
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def binaryTreePaths(self, root: t.Optional[TreeNode]) -> t.List[str]:
        path_list: t.List[str] = []
        self.binaryTreePathsHelper(root, path_list, "")
        return path_list
    
    def binaryTreePathsHelper(
        self
        , root: t.Optional[TreeNode]
        , path_list: t.List[str]
        , prev_path: str
    ) -> None:
        if not root:
            return
        
        if not prev_path:
            prev_path = f"{root.val}"
        else:
            prev_path = f"{prev_path}->{root.val}"
        
        if not root.left and not root.right:
            path_list.append(prev_path)
            return 
  
        self.binaryTreePathsHelper(root.left, path_list, prev_path)
        self.binaryTreePathsHelper(root.right, path_list, prev_path)

